packages/docusaurus-utils-validation/src/tags.ts

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import path from 'path';
+import fs from 'fs-extra';
+import _ from 'lodash';
+import Joi from 'joi';
+import YAML from 'js-yaml';
+import type {
+  Tag,
+  TagsFile,
+  TagsFileInput,
+  TagsPluginOptions,
+} from '@docusaurus/utils';
+
+const tagDefinitionSchema = Joi.object<TagsFile>().pattern(
+  Joi.string(),
+  Joi.object({
+    label: Joi.string(),
+    description: Joi.string(),
+    permalink: Joi.string(),
+  }).allow(null),
+);
+
+function validateDefinedTags(tags: unknown): Joi.ValidationResult<TagsFile> {
+  return tagDefinitionSchema.validate(tags);
+}
+
+export function ensureUniquePermalinks(tags: TagsFile): void {
+  const permalinks = new Set<string>();
+  const duplicates = new Set<string>();
+
+  for (const [, tag] of Object.entries(tags)) {
+    const {permalink} = tag;
+    if (permalinks.has(permalink)) {
+      duplicates.add(permalink);
+    } else {
+      permalinks.add(permalink);
+    }
+  }
+
+  if (duplicates.size > 0) {
+    throw new Error(
+      `Duplicate permalinks found: ${Array.from(duplicates).join(', ')}`,
```

Comment:
probably worth displaying this on multiple lines starting with `-` because some permalinks can be quite long or contain complex chars making things hard to read.

Also worth improving the error message because we don't even know it's about tags right now. The error message should mention it's about the tags file at least.